### ISSUE TITLE
The api 'tf.data.experimental.CsvDataset' performs very slow in test.

### ISSUE THREAD

**Issue by MrDavidG** [tensorflow/tensorflow#30041]
**System info.**
 - cuda: 10.1
 - python: 3.6.6
 - tensorflow: 1.13.1
 - gpu: Quadro P5000

**Test Data.**
Csv dataset with 430 columns (all in floats), the first 429 columns as features and the last column as labels. There are totally 1928 classes and 57909 instances. 

**Problem.**
I test the speed of the csv readers, api ```tf.data.experimental.CsvDataset``` and the common method ```tf.placeholder()```. Here is the code:

with ```tf.data.experimental.CsvDataset```:
```
def parse_data(x, n_classes):
    x = tf.convert_to_tensor(x)
    return x[:-1], tf.one_hot(indices=tf.cast(x[-1], tf.int32), depth=n_classes)

if __name__=='__main__':
    dataset_train = tf.data.experimental.CsvDataset('/home/david/Dataset/timit/test.csv', [tf.float32] * 430,
                                                    header=False,
                                                    field_delim=' ')
    dataset_train = dataset_train.map(lambda *x_: parse_data(x_, 1928))
    dataset_train = dataset_train.batch(128)
    dataset_train = dataset_train.prefetch(1)
    iterator = dataset_train.make_initializable_iterator()

    x_in, y = iterator.get_next()

    x = tf.layers.Dense(units=1024, activation=tf.nn.relu)(x_in)
    x = tf.layers.Dense(units=1024, activation=tf.nn.relu)(x)
    x = tf.layers.Dense(units=1024, activation=tf.nn.relu)(x)
    x = tf.layers.Dense(units=1024, activation=tf.nn.relu)(x)
    logits = tf.layers.Dense(units=1928, activation=None)(x)

    loss = tf.losses.softmax_cross_entropy(y, logits)
    optimizer = tf.train.AdamOptimizer()
    optimizer.minimize(loss)

    sess = tf.Session()
    sess.run(tf.global_variables_initializer())
    sess.run(iterator.initializer)
    running_loss = 0.0
    time_last = time.time()
    epoch = 0
    i = 0
    while True:
        try:
            running_loss += sess.run(loss)  # , feed_dict={x: data, y: labels})
            if (i + 1) % 5 == 0:
                print('\r[epoch: %2d, batch: %5d, time: %5f] loss: %.3f' % (
                    epoch + 1, i + 1, time.time() - time_last, running_loss / i), end=' ')
                time_last = time.time()
            i += 1
        except tf.errors.OutOfRangeError:
            pass
```

with ```tf.placeholder```:
```
if __name__ == '__main__':
    x_in = tf.placeholder(shape=[128, 429], dtype=tf.float32)
    y_in = tf.placeholder(shape=[128], dtype=tf.int32)
    y = tf.one_hot(y_in, depth=1928)

    x = tf.layers.Dense(units=1024, activation=tf.nn.relu)(x_in)
    x = tf.layers.Dense(units=1024, activation=tf.nn.relu)(x)
    x = tf.layers.Dense(units=1024, activation=tf.nn.relu)(x)
    x = tf.layers.Dense(units=1024, activation=tf.nn.relu)(x)
    logits = tf.layers.Dense(units=1928, activation=None)(x)

    loss = tf.losses.softmax_cross_entropy(y, logits)
    optimizer = tf.train.AdamOptimizer()
    optimizer.minimize(loss)

    sess = tf.Session()
    sess.run(tf.global_variables_initializer())

    w = pd.read_csv('/home/david/Dataset/timit/test.csv', header=None, delim_whitespace=True).values

    for epoch in range(23):
        running_loss = 0.0
        time_last = time.time()
        i = 0
        indexes = np.random.permutation(w.shape[0])
        w_ = w[indexes, :]
        while True:
            if i * 128 + 128 > w.shape[0]:
                break
            running_loss += sess.run(loss,
                                     feed_dict={x_in: w_[i * 128:i * 128 + 128, :-1],
                                                y_in: w_[i * 128:i * 128 + 128, -1]})
            if (i + 1) % 5 == 0:
                print('\r[epoch: %2d, batch: %5d, time: %5f] loss: %.3f' % (
                    epoch + 1, i + 1, time.time() - time_last, running_loss / i), end=' ')
                time_last = time.time()
            i += 1
```

**Result (Time for training five batches).**
 - ```tf.placeholder``` method: 0.013263s
 - ```tf.data.experimental.CsvDataset``` method: 1.382647s

**Problem.**
 - Api ```tf.data.experimental.CsvDataset``` is so slow in the above test. I guess it is partly because that ```tf.data.experimental.CsvDataset``` do io operations before each batch to extract data from csv file. Is this ture, or there are other reasons?

 - However, it is too slow comparing to ```tf.placeholder```. Is there any chance for improvement? How can I set the ```tf.data.experimental.CsvDataset ```api to load all csv data at the very beginning?

 - Or can I say that ```tf.data.experimental.CsvDataset``` is implemented only for the csv dataset that is too big to store in the memory? Because the time cost seems like intolerable.

**Comment by ravikyram** [comment]
Can you please help us with test.csv to reproduce the issue.Thanks.

**Comment by MrDavidG** [comment]
Sure, the ```test.csv``` can be downloaded [here](https://drive.google.com/file/d/1Lo0CULsDWt8qoFzFlq6FzCNb-Rt9bp9P/view?usp=sharing), but I think there is nothing special for this dataset.

**Comment by robieta** [comment]
All in all this isn't really surprising. There's a reason people don't use CSVs for high performance pipelines. That said, there are couple adjustments to make the comparison more apples to apples:

1) Drop the network and just run through the dataset in isolation. (Since that's all we're interested in.)
2) **Time the Pandas load**. This is the big one. Loading the CSV is by far the most expensive part of both pipelines, so it should be timed in both.
3) Run for more than 5 steps. You generally need to run for a few dozen to reasonably amortize and startup costs. 

When you correct for those factors the delta drops from ~100x to ~3.5x:
https://colab.sandbox.google.com/gist/robieta/8ac90d720108666ddd58cff9f1d2a88c/csv_profile.ipynb

Taking a step back, however, the truth is that if you want a very high performance pipeline on data that doesn't fit into memory tf.data offers a variety of solutions. (TFRecordDataset for parallel reads of TFRecords, FixedLengthRecordDataset if you just want to dump the raw NumPy buffer and then decode it, etc.) Ultimately an appropriately chosen binary format will always outperform a text based format; it's just that for a lot of applications CSV happens to be fast enough.

**Comment by Pearseak** [comment]
> All in all this isn't really surprising. There's a reason people don't use CSVs for high performance pipelines. That said, there are couple adjustments to make the comparison more apples to apples:
> 
> 1. Drop the network and just run through the dataset in isolation. (Since that's all we're interested in.)
> 2. **Time the Pandas load**. This is the big one. Loading the CSV is by far the most expensive part of both pipelines, so it should be timed in both.
> 3. Run for more than 5 steps. You generally need to run for a few dozen to reasonably amortize and startup costs.
> 
> When you correct for those factors the delta drops from ~100x to ~3.5x:
> https://colab.sandbox.google.com/gist/robieta/8ac90d720108666ddd58cff9f1d2a88c/csv_profile.ipynb
> 
> Taking a step back, however, the truth is that if you want a very high performance pipeline on data that doesn't fit into memory tf.data offers a variety of solutions. (TFRecordDataset for parallel reads of TFRecords, FixedLengthRecordDataset if you just want to dump the raw NumPy buffer and then decode it, etc.) Ultimately an appropriately chosen binary format will always outperform a text based format; it's just that for a lot of applications CSV happens to be fast enough.

I really want to use TFRecord but TFRecordWriter is superslow.
I already check that Example.SerializeToString() is the cause.
In my case in took 15secs to process 50000 records and I have 220,000,000 .......that's 5 days!!
I mean I prepare that 220m rows with C# in 15mimutes by Parallel.ForEach.
It would be nice if they can explain how to construct this TFRecord in binary stucture so that we can use any tool or language we want.

**Comment by robieta** [comment]
For advice on optimizing a particular implementation, I would suggest you consult [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since there a larger community that reads questions there. Github issues are more narrowly scoped to bugs and, in this case, feature requests. However I'll let @rachellim have the final word on that front. (I just happened to be passing through.)

**Comment by rachellim** [comment]
Thanks @robieta ! Looking at this [old github issue](https://github.com/tensorflow/tensorflow/issues/4467), the poor performance of the `SerializeToString` has been noticed before. Can you check what implementation of `protobuf` is being used?

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
